script/surface-connections.lua

```diff
@@ -0,0 +1,60 @@
+function ClearAllSurfaceConnections()
+  global.ConnectedSurfaces = {}
+end
+
+local function sorted_pair(number1, number2)
+  return (number1 < number2) and (number1..'|'..number2) or (number2..'|'..number1)
+end
+
+local function lazy_subtable(a_table, key)
+  local subtable = a_table[key]
+  if not subtable then
+    subtable = {}
+    a_table[key] = subtable
+  end
+  return subtable
+end
+
+function DisconnectSurfaces(entity1, entity2)
+  if not (entity1 and entity1.valid and entity2 and entity2.valid) then
+    return -- this gets automatically cleaned up in find_surface_connections()
+  end
+
+  local surface_pair_key = sorted_pair(entity1.surface.index, entity2.surface.index)
+  local surface_connections = global.ConnectedSurfaces[surface_pair_key]
+
+  if surface_connections then
+    surface_connections[sorted_pair(entity1.unit_number, entity2.unit_number)] = nil
+  end
+end
+
+function ConnectSurfaces(entity1, entity2, network_id)
+  if not (entity1 and entity1.valid and entity2 and entity2.valid) then
+    error("both entities must be valid")
+  end
+  if entity1.surface == entity2.surface then
+    error("connecting entities on the same surface is nonsensical")
```

Comment:
Errors should support language files

Reply:
I skip over invalid entities with either an chat error message or just a log notification.
Sadly users tend to report errors to the mod throwing them not the one causing them and I don't want to deal with those reports. ;)